filesink/filesink.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	lastBinding := -1
+	for it.Next() {
+		b := t.bindings[it.Binding]
+
+		if lastBinding != -1 && lastBinding != it.Binding {
+			if err := finishFile(); err != nil {
+				return nil, fmt.Errorf("finishFile after binding change: %w", err)
+			}
+		}
+		lastBinding = it.Binding
+
+		if encoder == nil {
+			startFile(b)
+		}
+
+		row := make([]any, 0, len(it.Key)+len(it.Values)+1)
+		row = append(row, it.Key.ToInterface()...)
+		row = append(row, it.Values.ToInterface()...)
+		if b.includeDoc {
+			row = append(row, it.RawJSON)
+		}
+
+		if err := encoder.Encode(row); err != nil {
+			return nil, fmt.Errorf("encoding row: %w", err)
+		}
+
+		if t.common.FileSizeLimit != 0 && encoder.Written() >= t.common.FileSizeLimit {
+			if err := finishFile(); err != nil {
+				return nil, fmt.Errorf("finishFile on file size limit: %w", err)
+			}
+		}
+
+	}
+	if err := it.Err(); err != nil {
+		return nil, fmt.Errorf("store iterator error: %w", err)
+	}
+
+	if err := finishFile(); err != nil {
+		return nil, fmt.Errorf("final finishFile: %w", err)
+	}
+
+	return func(ctx context.Context, _ *protocol.Checkpoint) (*pf.ConnectorState, m.OpFuture) {
+		checkpointJSON, err := json.Marshal(t.state)
+		if err != nil {
+			return nil, m.FinishedOperation(fmt.Errorf("creating checkpoint json: %w", err))
+		}
+
+		return &pf.ConnectorState{UpdatedJson: checkpointJSON}, nil
+	}, nil
+}
+
+// Acknowledge is a no-op since the files for this transaction were already uploaded to their
+// desired location in Store, and the driver checkpoint contains the last file counts used.
+func (t *transactor) Acknowledge(ctx context.Context) (*pf.ConnectorState, error) { return nil, nil }
+
+func (t *transactor) Destroy() {}
+
+// StdConstraints represents standard constraints for a projection when materializing columns of
+// data to files.
+func StdConstraints(p *pf.Projection) *pm.Response_Validated_Constraint {
+	_, isNumeric := boilerplate.AsFormattedNumeric(p)
+
+	var constraint = new(pm.Response_Validated_Constraint)
+	switch {
+	case p.IsPrimaryKey:
+		constraint.Type = pm.Response_Validated_Constraint_LOCATION_REQUIRED
+		constraint.Reason = "All Locations that are part of the collections key are required"
+	case p.IsRootDocumentProjection():
+		constraint.Type = pm.Response_Validated_Constraint_LOCATION_RECOMMENDED
+		constraint.Reason = "The root document should usually be materialized"
+	case len(p.Inference.Types) == 0 || slices.Equal(p.Inference.Types, []string{"null"}):
+		constraint.Type = pm.Response_Validated_Constraint_FIELD_FORBIDDEN
+		constraint.Reason = "Cannot materialize a field with no types"
+	case p.Field == "_meta/op":
+		constraint.Type = pm.Response_Validated_Constraint_LOCATION_RECOMMENDED
+		constraint.Reason = "The operation type should usually be materialized"
+	case strings.HasPrefix(p.Field, "_meta/"):
+		constraint.Type = pm.Response_Validated_Constraint_FIELD_OPTIONAL
+		constraint.Reason = "Metadata fields fields are able to be materialized"
```

Comment:
Do we have an established convention for not including _meta fields by default? I feel like we usually do include them as recommended (e.g. uuid and flow_published_at, etc.)